frame/nfts/src/features/atomic_swap.rs

```diff
@@ -0,0 +1,152 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::*;
+use frame_support::{
+	pallet_prelude::*,
+	traits::{Currency, ExistenceRequirement::KeepAlive},
+};
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	pub fn do_create_swap(
+		sender: T::AccountId,
+		collection_id: T::CollectionId,
+		item_id: T::ItemId,
+		desired_collection_id: T::CollectionId,
+		desired_item_id: T::ItemId,
+		maybe_price: Option<ItemPrice<T, I>>,
+		maybe_duration: Option<<T as SystemConfig>::BlockNumber>,
+	) -> DispatchResult {
+		let item = Item::<T, I>::get(&collection_id, &item_id).ok_or(Error::<T, I>::UnknownItem)?;
+		ensure!(item.owner == sender, Error::<T, I>::NoPermission);
+
+		Item::<T, I>::get(&desired_collection_id, &desired_item_id)
+			.ok_or(Error::<T, I>::UnknownItem)?;
+
+		let now = frame_system::Pallet::<T>::block_number();
+		let deadline = maybe_duration.map(|d| d.saturating_add(now));
+
+		PendingSwapOf::<T, I>::insert(
+			&collection_id,
+			&item_id,
+			PendingSwap {
+				desired_collection: desired_collection_id,
+				desired_item: desired_item_id,
+				price: maybe_price,
+				deadline,
+			},
+		);
+
+		Self::deposit_event(Event::SwapCreated {
+			collection: collection_id,
+			item: item_id,
+			desired_collection: desired_collection_id,
+			desired_item: desired_item_id,
+			price: maybe_price,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	pub fn do_cancel_swap(
+		sender: T::AccountId,
+		collection_id: T::CollectionId,
+		item_id: T::ItemId,
+	) -> DispatchResult {
+		let item = Item::<T, I>::get(&collection_id, &item_id).ok_or(Error::<T, I>::UnknownItem)?;
```

Comment:
If this `<T, i>` annoys you, you can add `#[pallet::getter(fn item)]` above the `Item` storage definition to use `Self::item(cid, iid)` instead, which is a bit shorter.

Reply:
> If this `Item::<T, I>::get` annoys you, you can add `#[pallet::getter(fn item)]` above the `Item` storage definition to use `Self::item(cid, iid)` instead, which is a bit shorter.

I personally don't see much difference here. Probably it even annoys me to see those hacky getters :) in other programming languages, you use getters to modify (or restrict) the output which makes more sence. E.g. `get_full_name { this.first_name + this.last_name }

Reply:
> Why is it important that the item exists here? Should it not be enough that the Swap exists? The item only has to exist if `!is_past_deadline`. Maybe worth it to put the seller into the Swap.

You're right! I moved it inside the `if !is_past_deadline {` block where it's actually needed